cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/preview/PreviewJsonSerDe.java

```diff
@@ -0,0 +1,85 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.store.preview;
+
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.format.io.JsonDecoder;
+import io.cdap.cdap.format.io.JsonEncoder;
+import io.cdap.cdap.format.io.JsonStructuredRecordDatumReader;
+import io.cdap.cdap.format.io.JsonStructuredRecordDatumWriter;
+import io.cdap.cdap.internal.io.SchemaTypeAdapter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.StringReader;
+import java.io.StringWriter;
+import java.lang.reflect.Type;
+
+/**
+ * Preview structured record serializer/deserializer.
+ */
+public class PreviewJsonSerDe implements JsonDeserializer<StructuredRecord>, JsonSerializer<StructuredRecord> {
+  private static final Logger LOG = LoggerFactory.getLogger(PreviewJsonSerDe.class);
+  private static final JsonStructuredRecordDatumReader JSON_DATUM_READER =
+    new JsonStructuredRecordDatumReader(false, true, true);
+  private static final JsonStructuredRecordDatumWriter JSON_DATUM_WRITER = new JsonStructuredRecordDatumWriter(true);
+
+  @Override
+  public StructuredRecord deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
+    throws JsonParseException {
+    JsonObject asJsonObject = json.getAsJsonObject();
+
+    JsonElement schema = asJsonObject.getAsJsonObject().get("schema");
+    JsonElement fields = asJsonObject.getAsJsonObject().get("fields");
+
+    Gson gson = new GsonBuilder().registerTypeAdapter(Schema.class, new SchemaTypeAdapter()).create();
+    Schema schemaToRead = gson.fromJson(schema, Schema.class);
+    try (JsonReader reader = new JsonReader(new StringReader(fields.getAsString()))) {
+      return JSON_DATUM_READER.read(new JsonDecoder(reader), schemaToRead);
+    } catch (IOException e) {
+      LOG.error("Error while deserializing structured record {}", e.getMessage(), e);
+    }
+    return null;
+  }
+
+  @Override
+  public JsonElement serialize(StructuredRecord src, Type typeOfSrc, JsonSerializationContext context) {
+    StringWriter strWriter = new StringWriter();
+    try (JsonWriter writer = new JsonWriter(strWriter)) {
+      JSON_DATUM_WRITER.encode(src, new JsonEncoder(writer));
```

Comment:
Why the serialize and deserialize are not symmetric?

Reply:
serialization will serialize logical types as string. For deserialize in preview mode, we would always want logical type fields as string. However, that does not comply with avro schema. Meaning, for non preview we would want structured records to contain original primitive types(not string) for logical type fields.